warehouse/migrations/env.py

```diff
@@ -42,37 +42,25 @@ def run_migrations_online():
     In this scenario we need to create an Engine
     and associate a connection with the context.
     """
-    connectable = context.config.attributes.get("connection", None)
+    options = context.config.get_section(context.config.config_ini_section)
+    url = options.pop("url")
+    connectable = create_engine(url, poolclass=pool.NullPool)
 
-    if connectable is None:
-        options = context.config.get_section(context.config.config_ini_section)
-        url = options.pop("url")
-        connectable = create_engine(url, poolclass=pool.NullPool)
+    with connectable.connect() as connection:
+        connection.execute(text("SET statement_timeout = 5000"))
+        connection.execute(text("SET lock_timeout = 4000"))
 
-        with connectable.connect() as connection:
-            connection.execute(text("SET statement_timeout = 5000"))
-            connection.execute(text("SET lock_timeout = 4000"))
-
-            context.configure(
-                connection=connection,
-                target_metadata=db.metadata,
-                compare_server_default=True,
-                transaction_per_migration=True,
-            )
-            with context.begin_transaction():
-                context.run_migrations()
-    else:
         context.configure(
-            connection=connectable,
+            connection=connection,
             target_metadata=db.metadata,
             compare_server_default=True,
             transaction_per_migration=True,
         )
-        context.execute(text("SET statement_timeout = 5000"))
-        context.execute(text("SET lock_timeout = 4000"))
-
         with context.begin_transaction():
+            connection.execute(text("SELECT pg_advisory_lock(hashtext('alembic'))"))
             context.run_migrations()
+            connection.execute(text("SELECT pg_advisory_unlock(hashtext('alembic'))"))
+            context.get_bind().commit()
```

Comment:
Should the commit happen before or after we release the lock?

Reply:
Well, we have to commit after we unlock, so the unlock happens.
Committing before kinda happens implicitly in the migrations that commit.

Maybe we want to commit *also* before running the migrations so the advisory lock is committed?

Reply:
I’d also be happy to proceed with this as is!

Reply:
I think this is actually backwards? We're not using the transaction based locks, so our locks transcend transactions either way, but by committing *after* we release the lock, we allow any migration process that is blocked on the lock to start happening prior to the final transaction being committed.

Since we run a transaction per migration.. I'm not sure what the actual implications of that is. It kind of depends on how the migration transactions are implemented (e.g. is each migration wrapped in a `BEGIN; COMMIT`, or does each migration start with a `COMMIT; BEGIN` and it relies on the final `COMMIT` for the last one?). It also may be a moot point because I think the first thing alembic is going to do is a `SELECT` against the alembic version table, which should end up blocked until the transactions commit anyways (since each transaction has to update that table to the new version).

I'm fine not changing it, but I *think* technically we want the release to happen after the commit, so just mentioning that in case @di or @miketheman want to change it.

Reply:
```suggestion
            connection.execute(text("SELECT pg_advisory_unlock(hashtext('alembic'))"))
            context.get_bind().commit()
```